### PR TITLE
Add Zipformer-MMI

### PR DESCRIPTION
**Pull request by yaozengwei**
This PR adds **Zipformer-MMI** recipe. 

It uses **CTC loss for warm-up** and then switches to MMI loss during training. 

For decoding, it uses HP (H is ctc_topo, P is token-level bi-gram) as decoding graph. Supported decoding methods are: 
- **1best**. Extract the best path from the decoding lattice as the decoding result.
- **nbest**. Extract n paths from the decoding lattice; the path with the highest score is the decoding result.
- **nbest-rescoring-LG**. Extract n paths from the decoding lattice, rescore them with an word-level 3-gram LM, the path with the highest score is the decoding result.
- **nbest-rescoring-3-gram**. Extract n paths from the decoding lattice, rescore them with an token-level 3-gram LM, the path with the highest score is the decoding result.
- **nbest-rescoring-4-gram**. Extract n paths from the decoding lattice, rescore them with an token-level 4-gram LM, the path with the highest score is the decoding result. 

Experimental results training on train-clean-100 (epoch-30-avg-10):
- 1best. 6.43 & 17.44
- nbest, nbest-scale=1.2, 6.43 & 17.45
- nbest-rescoring-LG, nbest-scale=1.2, 5.87 & 16.35
- nbest-rescoring-3-gram,  nbest-scale=1.2, 6.19 & 16.57
- nbest-rescoring-4-gram,  nbest-scale=1.2, 5.87 & 16.07

Experimental results training on full librispeech (epoch-30-avg-10):
- 1best. 2.54 & 5.65
- nbest, nbest-scale=1.2, 2.54 & 5.66
- nbest-rescoring-LG, nbest-scale=1.2, 2.49 & 5.42
- nbest-rescoring-3-gram,  nbest-scale=1.2, 2.52 & 5.62
- nbest-rescoring-4-gram,  nbest-scale=1.2, 2.5 & 5.51

### CODE REVIEW

File: .github/scripts/run-librispeech-zipformer-mmi-2022-12-08.sh [hunk 1]
```diff
@@ -0,0 +1,96 @@
...
+  local fname=${BASH_SOURCE[1]##*/}
+  echo -e "$(date '+%Y-%m-%d %H:%M:%S') (${fname}:${BASH_LINENO[0]}:${FUNCNAME[1]}) $*"
+}
+
+cd egs/librispeech/ASR
+
+repo_url=https://huggingface.co/Zengwei/icefall-asr-librispeech-zipformer-mmi-2022-12-08
+
+log "Downloading pre-trained model from $repo_url"
+git lfs install
```
Comment (csukuangfj) [inline .github/scripts/run-librispeech-zipformer-mmi-2022-12-08.sh:16]:
```suggestion
```

We are using 
```
GIT_LFS_SKIP_SMUDGE=1 git clone $repo_url
```
and
```
git lfs pull --include "data/*"
git lfs pull --include "exp/cpu_jit.pt"
git lfs pull --include "exp/pretrained.pt"
```
so it is fine to remove `git lfs install`.

File: .github/scripts/run-librispeech-zipformer-mmi-2022-12-08.sh [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+GIT_LFS_SKIP_SMUDGE=1 git clone $repo_url
+repo=$(basename $repo_url)
+
+log "Display test files"
+tree $repo/
+soxi $repo/test_wavs/*.wav
+ls -lh $repo/test_wavs/*.wav
+
+pushd $repo/exp
+git lfs pull --include "data/*"
```
Comment (csukuangfj) [inline .github/scripts/run-librispeech-zipformer-mmi-2022-12-08.sh:26]: I suggest that you only pull files that are actually needed.

File: .github/workflows/run-librispeech-2022-12-08-zipformer-mmi.yml [hunk 3]
```diff
@@ -0,0 +1,167 @@
+# Copyright      2022  Fangjun Kuang (csukuangfj@gmail.com)
```
Comment (csukuangfj) [inline .github/workflows/run-librispeech-2022-12-08-zipformer-mmi.yml:1]:
```suggestion
# Copyright      2022  Zengwei Yao
```

File: .github/scripts/run-librispeech-zipformer-mmi-2022-12-08.sh (hunk 1 above)
Comment (yaozengwei) [reply .github/scripts/run-librispeech-zipformer-mmi-2022-12-08.sh:16]: OK. Thanks.

File: egs/librispeech/ASR/RESULTS.md [hunk 4]
```diff
@@ -1,5 +1,62 @@
...
+  --use-fp16 1 \
+  --exp-dir zipformer_mmi/exp
+```
+
+The decoding commands for the transducer branch are:
+```bash
+export CUDA_VISIBLE_DEVICES="5"
+
+for m in nbest nbest-rescoring-LG nbest-rescoring-3-gram nbest-rescoring-4-gram; do
+  ./zipformer_mmi/decode.py \
```
Comment (csukuangfj) [inline egs/librispeech/ASR/RESULTS.md:47]: Does it support `--use-averaged-model=true`?

File: egs/librispeech/ASR/zipformer_mmi/jit_pretrained.py [hunk 5]
```diff
@@ -0,0 +1,391 @@
...
+        Used only when "method" is one of the following values:
+        nbest, nbest-rescoring, and nbest-oracle
+        A smaller value results in more unique paths.
+        """,
+    )
+
+    parser.add_argument(
+        "--ngram-lm-scale",
+        type=float,
+        default=0.1,
```
Comment (csukuangfj) [inline egs/librispeech/ASR/zipformer_mmi/jit_pretrained.py:178]: Could you use the same default value as train.py?

File: egs/librispeech/ASR/zipformer_mmi/jit_pretrained.py [hunk 6]
```diff
@@ -0,0 +1,391 @@
...
+          the highest score is the decoding result.
+        """,
+    )
+
+    parser.add_argument(
+        "--sample-rate",
+        type=int,
+        default=16000,
+        help="The sample rate of the input sound file",
+    )
```
Comment (csukuangfj) [inline egs/librispeech/ASR/zipformer_mmi/jit_pretrained.py:144]: Looks like it is never used.

File: egs/librispeech/ASR/zipformer_mmi/model.py [hunk 7]
```diff
@@ -0,0 +1,75 @@
+# Copyright    2021  Xiaomi Corp.        (authors: Fangjun Kuang, Wei Kang)
```
Comment (csukuangfj) [inline egs/librispeech/ASR/zipformer_mmi/model.py:1]:
```suggestion
# Copyright    2022  Xiaomi Corp.        (authors: Zengwei Yao)
```

Comment (yaozengwei) [reply egs/librispeech/ASR/zipformer_mmi/model.py:1]: Thanks a lot.

File: egs/librispeech/ASR/zipformer_mmi/jit_pretrained.py (hunk 5 above)
Comment (yaozengwei) [reply egs/librispeech/ASR/zipformer_mmi/jit_pretrained.py:178]: It is the `lm_scale`, which is searched out with decode.py.

**Review by csukuangfj — APPROVED** [review]

**Comment by wangtiance** [comment]
> It uses CTC loss for warm-up and then switches to MMI loss during training.

Is it because the model won't converge without warm-up? How does the WER compare to CTC training only? Thanks.

File: egs/librispeech/ASR/zipformer_mmi/decode.py [hunk 8]
```diff
@@ -0,0 +1,736 @@
...
+      dl:
+        PyTorch's dataloader containing the dataset to decode.
+      params:
+        It is returned by :func:`get_params`.
+      model:
+        The neural model.
+      HP:
+        The decoding graph. H is ctc_topo, P is token-level bi-gram LM.
+      bpe_model:
+        The BPE model. Used only when params.decoding_method is ctc-decoding.
```
[the remark this answers is not in the document]
Comment (umbertocappellazzo) [reply egs/librispeech/ASR/zipformer_mmi/decode.py]: Hi! Is ctc-decoding supported as of now?

**Pull request merged**